stdlib/immutablearray.gr

```diff
@@ -0,0 +1,921 @@
+ *
+ * @example zipWith((a, b) => a + b, fromList([1, 2, 3]), fromList([4, 5, 6])) == fromList([5, 7, 9])
+ * @example zipWith((a, b) => a * b, fromList([1, 2, 3]), fromList([4, 5])) == fromList([4, 10])
+ * 
+ * @since v0.6.0
+ */
+export let zipWith = (fn, array1, array2) => {
+  fromList(List.zipWith(fn, toList(array1), toList(array2)))
+}
+
+/**
+ * Produces two arrays by splitting apart an array of tuples.
+ *
+ * @param array: The array of tuples to split
+ * @returns An array containing all elements from the first tuple element, and an array containing all elements from the second tuple element
+ *
+ * @since v0.6.0
+ */
+export let unzip = array => {
+  let (list1, list2) = List.unzip(toList(array))
+  (fromList(list1), fromList(list2))
+}
+
+/**
+ * Concatenates an array of strings into a single string, separated by a separator string.
+ *
+ * @param separator: The separator to insert between items in the string
+ * @param array: The input strings
+ * @returns The concatenated string
+ *
+ * @since v0.6.0
+ */
+export let join = (separator, array) => {
+  let buf = Buffer.make(16)
+  let mut i = 0
+  forEach(x => {
+    let toAdd = if (i != array.length - 1) x ++ separator else x
+    i += 1
+    Buffer.addString(toAdd, buf)
+  }, array)
+  Buffer.toString(buf)
+}
+
+let clampIndex = (len, index) => {
+  Number.min(len, Number.max(0, wrapNegativeIndex(len, index)))
+}
+
+/**
+ * Slices an array given zero-based start and end indexes. The value
+ * at the end index will not be included in the result.
+ *
+ * If either index is a negative number, it will be treated as a reverse index from
+ * the end of the array.
+ *
+ * @param start: The index of the array where the slice will begin (inclusive)
+ * @param end: The index of the array where the slice will end (exclusive)
+ * @param array: The array to be sliced
+ * @returns The subset of the array that was sliced
+ * 
+ * @example slice(0, 2, fromList(['a', 'b', 'c'])) == fromList(['a', 'b'])
+ * @example slice(1, -1, fromList(['a', 'b', 'c'])) == fromList(['b'])
+ *
+ * @since v0.6.0
+ */
+export let slice = (start, end, array) => {
+  let begin = clampIndex(array.length, start)
+  let end = clampIndex(array.length, end)
+  let mut i = array.length
+  fromList(
+    reduceRight((x, acc) => {
+      i -= 1
+      if (i >= begin && i < end) [x, ...acc] else acc
+    }, [], array)
+  )
+}
+
+/**
+ * Sorts the given array based on a given comparator function.
+ *
+ * Ordering is calculated using a comparator function which takes two array elements and must return 0 if both are equal, a positive number if the first is greater, and a negative number if the first is smaller.
+ * @param comp: The comparator function used to indicate sort order
+ * @param array: The array to be sorted
+ * @returns The sorted array
+ *
+ * @since v0.6.0
+ */
+export let sort = (comp, array) => {
+  fromList(List.sort(comp, toList(array)))
+}
+
+/**
+ * Rotates array elements by the specified amount to the left.
```

Comment:
Double checking that this is "correct" (in light of list and array operating in the opposite directions).

Reply:
In my opinion this seems like the more appropriate direction to do a rotation in so I just went with a left rotation, although #1470 hasn't been commented on yet so I don't know what the consensus on what the "correct" behavior is